### PR TITLE
Implement Login via WebView

### PR DESCRIPTION
**Pull request by charlag**
I've tested all the scenarios I can think of but I would appreciate more testing.

### CODE REVIEW

**Comment by connyduck** [comment]
closes https://github.com/tuskyapp/Tusky/issues/2134
closes https://github.com/tuskyapp/Tusky/issues/2040

File: app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val data = OauthLogin.parseData(intent)
+
+        setContentView(binding.root)
+
+        setSupportActionBar(binding.loginToolbar)
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        supportActionBar?.setDisplayShowTitleEnabled(false)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt:81]: I'd give it a title that says "Login"

File: app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        val webView = binding.loginWebView
+        webView.settings.allowContentAccess = false
+        webView.settings.allowFileAccess = false
+        webView.settings.databaseEnabled = false
+        webView.settings.displayZoomControls = false
+        webView.settings.javaScriptCanOpenWindowsAutomatically = false
+        webView.settings.userAgentString += "Tusky ${BuildConfig.VERSION_NAME}"
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt:89]: maybe add the Android Version like here https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt#L81

File: app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt [hunk 3]
```diff
@@ -137,130 +159,92 @@ class LoginActivity : BaseActivity(), Injectable {
...
-
-                redirectUserToAuthorizeAndLogin(domain, clientId)
-            }
+        setLoading(true)
 
-            override fun onFailure(call: Call<AppCredentials>, t: Throwable) {
+        lifecycleScope.launch {
+            val credentials: AppCredentials = try {
+                mastodonApi.authenticateApp(
+                    domain, getString(R.string.app_name), oauthRedirectUri.toString(),
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt:167]: `oauthRedirectUri` already is a String
(also 3 more times below)

File: app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt [hunk 4]
```diff
@@ -137,130 +159,92 @@ class LoginActivity : BaseActivity(), Injectable {
...
+            .scheme("https")
+            .host(domain)
+            .addPathSegments(MastodonApi.ENDPOINT_AUTHORIZE)
+            .addQueryParameter("client_id", clientId)
+            .addQueryParameter("redirect_uri", oauthRedirectUri.toString())
+            .addQueryParameter("response_type", "code")
+            .addQueryParameter("scope", OAUTH_SCOPES)
+            .build()
+        val uri = Uri.parse(url.toString())
+        doWebViewAuth.launch(LoginData(uri, Uri.parse(oauthRedirectUri.toString())))
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt:208]: `oauthRedirectUri.toUri()` 
https://developer.android.com/reference/kotlin/androidx/core/net/package-summary#(kotlin.String).toUri()

**Review by connyduck — COMMENTED** [review]
Works very well 😍

**Comment by connyduck** [comment]
Oh, and can you please move all Login things into the `components.login` package?

File: app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt (hunk 2 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt:89]: Good call!

File: app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt (hunk 3 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/LoginActivity.kt:167]: Oh yeah I made it a `HttpUrl` but it's not actually a "http" URL so I rolled it back.

File: app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt (hunk 1 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt:81]: We only have `action_login` I'm a bit hesitant to add a string just for that. If you say yes, I will do it

File: app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt (hunk 2 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/LoginWebViewActivity.kt:89]: I now injected exactly the same UA but then
 - Default UA has a lot of useful info e.g. engine version
 - Default UA already has Android version in it
 - We don't need OkHttp version for this UA

But I've changed a format now to be closer to what UA should be

**Comment by connyduck** [comment]
You need to fix the code formatting so the CI check passes, looks good otherwise

**Comment by charlag** [comment]
@connyduck thanks, I missed that. Can't squash commits nicer now because there's a move to package inbetween but eh

**Pull request merged**